File: sendSMS.py
#!/usr/bin/python2.7 

# File name: sendSMS.py
# Author: Daniel Gullin
# Date created: 2016-10-19
# Version: 0.2
# License: GPL

import logging
import requests
import argparse
import re

### Settings ###
# File with msisdn
file = "phoneNumbers.cfg"

# If no source msisdn we use xxx as default (max 11 car)
default = "xxx"

# Base64-encoded authorization (https://www.base64encode.org/)
# We use user:PASSWORD
auth = "dXNlcjpQQVNTV09SRA=="

# The URL to GenericSMS HTTP API
url = "https://api.genericmobile.se/SmsGateway/api/v1/Message"

# Log SMS (/var/log/sendSMS.log)
log = True

###################
# DON'T TOUCH BELOW
###################

if log:
        logging.basicConfig(level=logging.DEBUG,
                format='%(asctime)s %(levelname)-8s %(message)s',
                datefmt='%b %d %Y %H:%M:%S',
                filename='/var/log/sendSMS.log',
                filemode='a')

# Arguments
parser = argparse.ArgumentParser(description="Send SMS through GenericSMS, usually used by OP5")
parser.add_argument("-s", "--source", help="Source MSISDN, default is BallouDrift")
parser.add_argument("-d", "--destination", type=str, help="Destination MSISDN, use international style +4670xxxxxxx format. If no destination is used file test.txt will be used")
parser.add_argument("-m", "--message", type=str, help="The Hot Message", required=True)
parser.add_argument("-v", "--verbose", help="Verbose output", action="store_true")
args = parser.parse_args()

# If no source number we shall use BallouDrift as default
if args.source == None: args.source = default

# The mighty send function!
def send_sms(msisdn):
        msisdn = re.sub(r'^0', "+46", msisdn)
        payload = '{\n    \"From\":\"%s\",\n    \"To\":["%s"],\n    \"Text\":\"%s"\n}' % (args.source, msisdn, args.message)
        headers = {
                'authorization': "Basic %s" % (auth),
                'accept': "application/json",
                'content-type': "application/json",
                'cache-control': "no-cache",
                'postman-token': "c7577af4-733d-fe98-ba1b-1728df06d687"
                }
        response = requests.request("POST", url, data=payload, headers=headers)
        if args.verbose: print "%s \n\n%s \n" % (payload, response.text)

        if log:
                logging.debug('From: %s', args.source)
                logging.debug('To: %s', msisdn )
                logging.debug("Text: %s", args.message )
                logging.debug("Response: %s", response.text )

# If no destination, use $phoneNumbers.cfg 
if args.destination == None:
        f = open(file, "r")
        for line in f:
                line = line.rstrip("\r\n")
                if line.isdigit(): send_sms(line)
else:
                send_sms(args.destination)

File: sendSMS.pl
#!/usr/bin/perl -w
use strict;
use HTTP::Lite;
use Getopt::Long;
#
# sms.pl - send sms through cellsynt
#
# 18 Jan 2007  Daniel Gullin   Created.
#
# Usage: sms.pl
#

# Change me?
my $username    = 'xxx';
my $passwd      = 'yyy';
my $sender      = 'zzz';
my $phonelist   = 'phoneNumbers.cfg';


#####################
### DON'T TOUCH BELOW
#####################
my $ref         = 'xxx';
my $url         = 'sms2.ballou.se';
my $http;
my $req;
my $message     = '0';
my @list        = ();
my $msisdn      = '0';
my $help        = '0';


GetOptions ('d=s' => \$msisdn,
            'm=s' => \$message,
            's=s' => \$sender,
            'h' => \$help);

# Help menu
sub help () {
        print "sms.pl: -[h] -s [msisdn] -d [msisdn] -m [message] \n\n";
        print "This script sends sms through sms engine. Either you can specify a single\n";
        print "msisdn or use a file ($phonelist) with multiple msisdn.\n\n";
        print "        -s       sender msisdn\n";
        print "        -d       destination msisdn\n";
        print "        -m       message\n";
        print "        -h       help menu (this)\n\n";
        exit
}

# Some dummy tests...
if ($help == 1) {
        help();
}
if ($message eq 0 || $sender eq 0) {
        help();
}

# First we nedd url encode the message
$message =~ s/([^A-Za-z0-9])/sprintf("%%%02X", ord($1))/seg;
$sender =~ s/([^A-Za-z0-9])/sprintf("%%%02X", ord($1))/seg;


# Here we go...
if ($msisdn > 0) {
        push(@list, "$msisdn");
} else {
        # Read the MSISDN
        open (FILE, $phonelist)         or die "couldn't open the phonelist file!";
        while (<FILE>) {
                chomp $_;
                $_ =~ s/^[+]46/0/g;
                push(@list, "$_") unless m/^#/;
        }
        close(FILE);
}

# The mighty send engine... ;)
foreach (@list) {
        $http = new HTTP::Lite;
                $req = $http->request("http://$url/http/get/SendSms.php?UN=$username&PW=$passwd&CR=$ref&O=$sender&D=$_&M=$message")
                or die "Unable dest get document: $!";
}

# Free some memory
undef @list;
undef $message;
undef $msisdn;
undef $help;

# Exit
exit
